File: 11week/day2/cokemania2/176962.py
# https://school.programmers.co.kr/learn/courses/30/lessons/176962
# 과제 진행하기

def time_to_minute(time):
    hour, minute = time.split(':')
    return int(hour) * 60 + int(minute)

def do_remain_plan(done, remain, remain_time):
    while len(remain) and remain_time:
        remain_plan = remain.pop()
        if remain_plan[2] <= remain_time:
            done.append(remain_plan)
            remain_time -= remain_plan[2]
        else:
            remain_plan[2] -= remain_time
            remain.append(remain_plan)
            remain_time = 0

def do_plan(i, plans, done, remain):
    if i == len(plans) - 1:
        done.append(plans[i])
        do_remain_plan(done, remain, 60 * 60 * 100 * 24)
        return 
    
    now_plan = plans[i]
    next_plan = plans[i + 1]
    
    end_time = time_to_minute(now_plan[1]) + int(now_plan[2])
    next_plan_start_time = time_to_minute(next_plan[1])

    if end_time > next_plan_start_time:
        now_plan[2] = end_time - next_plan_start_time
        remain.append(now_plan)
    else:
        done.append(now_plan)
        remain_time = next_plan_start_time - end_time
        do_remain_plan(done, remain, remain_time)


def solution(plans):
    done = []
    remain = []
    plans.sort(key = lambda x : x[1])
    for i in range(len(plans)):
        do_plan(i, plans, done, remain)
    
    return [d[0] for d in done]

File: 11week/corinthionia/176962.js
const timeToNumber = (time) => {
    const t = time.split(':').map(Number);
    return t[0] * 60 + t[1];
}

function solution(plans) {
    const times = plans.map((p) => {
        return [p[0], timeToNumber(p[1]), Number(p[2])];
    });
    
    times.sort((a, b) => a[1] - b[1]);
    
    const stk = [];
    const answer = [];
    
    for (let i = 0; i < times.length-1; i++) {
        stk.push([times[i][0], times[i][2]]);
        let gap = times[i+1][1] - times[i][1]; 
        
        while (0 < gap && stk.length) {

            const [tn, td] = stk.pop();
            if (td <= gap) { // 해당 과제 끝내기 가능
                answer.push(tn);
                gap -= td;
            } else {
                stk.push([tn, td-gap]);
                gap = 0;
            }
        }
    }
    
    answer.push(times[times.length-1][0]);
    return [...answer, ...stk.reverse().map((n) => n[0])]
}